mission-7/queries.sql

```diff
@@ -0,0 +1,126 @@
+/*
+  다음 경우들에 대해 총 14개의 SQL 쿼리를 작성해 주세요.
+  예시로 값이 필요한 경우 적당한 값으로 채워넣어서 작성하면 됩니다. 
+*/
+
+/*
+  1. 내 정보 업데이트 하기
+  - 닉네임을 "test"로 업데이트
+  - 현재 로그인한 유저 id가 1이라고 가정
+*/
+-- INSERT INTO "User" (email, nickname, password) VALUES ('test@test.com', 'user1', '12345678a');
+UPDATE "User" SET nickname = 'test' WHERE id = 1;
+-- SELECT * from "User";
+
+/*
+  2. 내가 생성한 상품 조회
+  - 현재 로그인한 유저 id가 1이라고 가정
+  - 최신 순으로 정렬
+  - 10개씩 페이지네이션, 3번째 페이지
+*/
+-- INSERT INTO "Product" (name, description, price, tags, "userId") VALUES ('test 상품', '테스트입니다.', 10000, ARRAY['test1', 'test2'], 2);
+SELECT * FROM "Product" WHERE "userId" = 1 LIMIT 10 OFFSET 20;
+-- SELECT * from "Product";
+
+/*
+  3. 내가 생성한 상품의 총 개수
+  - 현재 로그인한 유저 id가 1이라고 가정
+*/
+SELECT COUNT(*) AS "총 개수" FROM "Product" WHERE "userId" = 1; 
+
+/*
+  4. 내가 좋아요 누른 상품 조회
+  - 현재 로그인한 유저 id가 1이라고 가정
+  - 최신 순으로 정렬
+  - 10개씩 페이지네이션, 3번째 페이지
+*/
+SELECT p.* 
+FROM "Product" as p
+JOIN "ProductLike" as pl
+on p.id = pl."productId"
+WHERE pl."userId" = 1
+ORDER BY pl."createdAt" DESC
+LIMIT 10
+OFFSET 20;
+
+/*
+  5. 내가 좋아요 누른 상품의 총 개수
+  - 현재 로그인한 유저 id가 1이라고 가정
+*/
+SELECT COUNT(p.*) 
+FROM "Product" as p
+JOIN "ProductLike" as pl
+on p.id = pl."productId"
+WHERE pl."userId" = 1;
+
+/*
+  6. 상품 생성
+  - 현재 로그인한 유저 id가 1이라고 가정
+*/
+INSERT INTO "Product" (name, description, price, tags, "userId") VALUES ('테스트 상품', '테스트입니다.', 50000, ARRAY['test1', 'test2'], 1);
+
+
+/*
+  7. 상품 목록 조회
+  - 상품명에 "test"가 포함된 상품 검색
+  - 최신 순으로 정렬
+  - 10개씩 페이지네이션, 1번째 페이지
+  - 각 상품의 좋아요 개수를 포함해서 조회하기
+*/
+SELECT p.*, count(pl."productId") as "좋아요 수"
+FROM "Product" as p
+LEFT JOIN "ProductLike" as pl
+ON p.id = pl."productId"
+WHERE p.name Like '%test%'
+GROUP BY p.id
+ORDER BY p."createdAt" DESC
+LIMIT 10;
```

Comment:
1. 좋아요를 서브쿼리에서 먼저 그룹핑해서, 중복 행을 조인 전에 축소한다면 조인 비용과 정렬, LIMIT에 필요한 모든 양을 최소화시킬 수 있어요.
2. 최신순 조회에서는 시간 순 뒤에 동률 시 안정적인 순서 보장을 위해 서브 필드를 사용하시면 좋습니다.
3. 만약 대소문자를 무시해야한다면 LIKE > ILIKE

```suggestion
SELECT p.*, COALESCE(l.like_cnt, 0) AS "좋아요 수"
FROM "Product" AS p
LEFT JOIN (
  SELECT "productId", COUNT(*) AS like_cnt
  FROM "ProductLike"
  GROUP BY "productId"
) AS l
  ON l."productId" = p.id
WHERE p.name LIKE '%test%'
ORDER BY p."createdAt" DESC, p.id DESC
LIMIT 10;
```